libs/use-viewer/src/index.ts

```diff
@@ -89,127 +39,186 @@ const useViewer = ({ streamName, streamAccountId, subscriberToken, handleError }
+          }, 1000);
+        } catch (error) {
+          handleInternalError(error);
         }
         break;
-      case 'inactive':
-        {
-          const source = event.data as MediaStreamSource;
-          if (!source.sourceId) source.sourceId = missedSourceId;
-          dispatch({ type: ViewerActionType.REMOVE_SOURCE, sourceId: source.sourceId, viewer: viewer.current });
+
+      case 'inactive': {
+        const remoteTrackSource = remoteTrackSources.get(sourceId);
+
+        if (remoteTrackSource) {
+          unprojectRemoteTrackSource(viewer, remoteTrackSource);
+
+          dispatch({ type: ViewerActionType.REMOVE_SOURCE, sourceId });
         }
+
         break;
+      }
+
       case 'viewercount':
         setViewerCount((event.data as ViewerCount).viewercount);
         break;
+
       case 'layers': {
         dispatch({
           medias: (event.data as MediaStreamLayers).medias,
-          type: ViewerActionType.UPDATE_SOURCES_QUALITIES,
-          viewer: viewer.current as View,
+          type: ViewerActionType.UPDATE_SOURCES_QUALITY_OPTIONS,
         });
         break;
       }
     }
-  }, []);
+    return;
+  };
 
-  const connect = async () => {
-    if (!viewer.current) return;
-    console.log('connecting');
-    try {
-      await viewer.current.connect({ events: ['active', 'inactive', 'layers', 'viewercount'] });
-    } catch (error) {
-      console.error(error);
-      viewer.current?.reconnect();
-    } finally {
-      console.log('register broadcastEvent');
-      viewer.current?.webRTCPeer?.initStats();
-      viewer.current?.webRTCPeer?.on('stats', (statistics: StreamStats) => {
-        dispatch({
-          statistics,
-          type: ViewerActionType.UPDATE_SOURCES_STATISTICS,
-          viewer: viewer.current as View,
-        });
-      });
+  const handleStats = (statistics: StreamStats) => {
+    dispatch({
+      statistics,
+      type: ViewerActionType.UPDATE_SOURCES_STATISTICS,
+    });
+  };
+
+  // Get the audio/video media IDs for the main stream from the track event
+  // and use them to create the main stream mapping
+  const handleTrack = (event: RTCTrackEvent) => {
+    const {
+      streams,
+      track: { kind },
+      transceiver: { mid },
+    } = event;
+
+    if (streams.length && mid !== null) {
+      const newMapping = { media: kind, mediaId: mid, trackId: kind };
+
+      if (kind === 'audio') {
+        setMainStreamAudioMapping(newMapping);
+      } else if (kind === 'video') {
+        setMainStreamVideoMapping(newMapping);
+      }
     }
   };
 
+  const projectToMainStream = async (sourceId: SourceId, prevSourceId?: SourceId) => {
+    const { current: viewer } = viewerRef;
+
+    if (!viewer) {
+      return;
+    }
+
+    if (prevSourceId) {
+      const prevRemoteTrackSource = remoteTrackSources.get(prevSourceId);
+
+      if (prevRemoteTrackSource) {
+        await reprojectToOriginalMapping(viewer, prevRemoteTrackSource);
```

Comment:
We don't need to project the previous track to its stream again, because it's already been projected.